File: bidragsevne/src/main/java/no/nav/bidrag/beregn/bidragsevne/periode/Periodiserer.java
package no.nav.bidrag.beregn.bidragsevne.periode;

import static java.util.stream.Collectors.toList;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

public class Periodiserer {

  private final Set<LocalDate> bruddpunkter = new HashSet<>();
  private boolean aapenSluttdato = false;

  public Periodiserer addBruddpunkt(LocalDate dato) {
    bruddpunkter.add(dato);
    return this;
  }

  private Periodiserer addBruddpunkter(Periode periode) {
    addBruddpunkt(periode.getDatoFra());
    if (periode.getDatoTil() == null) {
      aapenSluttdato = true;
    } else {
      addBruddpunkt(periode.getDatoTil());
    }
    return this;
  }

  public Periodiserer addBruddpunkter(PeriodisertGrunnlag grunnlag) {
    addBruddpunkter(grunnlag.getDatoFraTil());
    return this;
  }

  Periodiserer addBruddpunkter(Iterable<? extends PeriodisertGrunnlag> grunnlagListe) {
    for (PeriodisertGrunnlag grunnlag : grunnlagListe) {
      addBruddpunkter(grunnlag);
    }
    return this;
  }

  public List<Periode> finnPerioder(LocalDate beregnDatoFra, LocalDate beregnDatoTil) {
    var sortertBruddpunktListe = bruddpunkter.stream().filter((dato) -> dato.isAfter(beregnDatoFra.minusDays(1)))
        .filter((dato) -> dato.isBefore(beregnDatoTil.plusDays(1))).sorted().collect(toList());

    List<Periode> perioder = new ArrayList<>();
    Iterator<LocalDate> bruddpunktIt = sortertBruddpunktListe.iterator();
    if (bruddpunktIt.hasNext()) {
      LocalDate start = bruddpunktIt.next();
      while (bruddpunktIt.hasNext()) {
        LocalDate end = bruddpunktIt.next();
        perioder.add(new Periode(start, end));
        start = end;
      }
      if (aapenSluttdato) {
        perioder.add(new Periode(start, null));
      }
    }
    return perioder;
  }
}

File: bidrag-beregn-core/src/main/kotlin/no/nav/bidrag/beregn/felles/periode/Periodiserer.kt
package no.nav.bidrag.beregn.felles.periode

import no.nav.bidrag.beregn.felles.bo.Periode
import no.nav.bidrag.beregn.felles.bo.PeriodisertGrunnlag
import java.time.LocalDate
import java.util.stream.Collectors

class Periodiserer {
    private val bruddpunkter: MutableSet<LocalDate> = HashSet()
    private var aapenSluttdato = false

    fun addBruddpunkt(dato: LocalDate): Periodiserer {
        bruddpunkter.add(dato)
        return this
    }

    private fun addBruddpunkter(periode: Periode) {
        addBruddpunkt(periode.datoFom)

        if (periode.datoTil == null) {
            aapenSluttdato = true
        } else {
            addBruddpunkt(periode.datoTil)
        }
    }

    fun addBruddpunkter(grunnlag: PeriodisertGrunnlag): Periodiserer {
        addBruddpunkter(grunnlag.getPeriode())
        return this
    }

    fun addBruddpunkter(grunnlagListe: Iterable<PeriodisertGrunnlag>): Periodiserer {
        for (grunnlag in grunnlagListe) {
            addBruddpunkter(grunnlag)
        }

        return this
    }

    // Setter perioder basert på fra- og til-dato
    fun finnPerioder(
        beregnDatoFom: LocalDate,
        beregnDatoTil: LocalDate,
    ): List<Periode> {
        val sortertBruddpunktListe =
            bruddpunkter.stream()
                .filter { dato: LocalDate -> dato.isAfter(beregnDatoFom.minusDays(1)) }
                .filter { dato: LocalDate -> dato.isBefore(beregnDatoTil.plusDays(1)) }
                .sorted().collect(Collectors.toList())

        val perioder: MutableList<Periode> = ArrayList()
        val bruddpunktIt = sortertBruddpunktListe.iterator()

        if (bruddpunktIt.hasNext()) {
            var start: LocalDate? = bruddpunktIt.next()

            while (bruddpunktIt.hasNext()) {
                val end = bruddpunktIt.next()
                perioder.add(Periode(start!!, end))
                start = end
            }

            if (aapenSluttdato) {
                perioder.add(Periode(start!!, null))
            }
        }

        return perioder
    }
}
